resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java

```diff
@@ -128,20 +129,34 @@ public long reservePermission(final int permits) {
 
         boolean canAcquireImmediately = modifiedState.nanosToWait <= 0;
         if (canAcquireImmediately) {
-            publishRateLimiterEvent(true, permits);
+            publishRateLimiterAcquisitionEvent(true, permits);
             return 0;
         }
 
         boolean canAcquireInTime = timeoutInNanos >= modifiedState.nanosToWait;
         if (canAcquireInTime) {
-            publishRateLimiterEvent(true, permits);
+            publishRateLimiterAcquisitionEvent(true, permits);
             return modifiedState.nanosToWait;
         }
 
-        publishRateLimiterEvent(false, permits);
+        publishRateLimiterAcquisitionEvent(false, permits);
         return -1;
     }
 
+    @Override
+    public void drainPermissions() {
```

Comment:
Please don't change state directly, but use the same approach as in `updateStateWithBackOff` https://github.com/resilience4j/resilience4j/blob/a2534c515eeb9361b755bbaab2210990f99ac123/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java#L157

Reply:
I don't think that is a good idea

to drain the number of permits to zero precisely in a thread safe manner I have to read the number of permits left and subtract that amount in one atomic operation

`updateStateWithBackOff` doesn't do that - explain why using it would be better?

Reply:
`updateStateWithBackOff` uses `compareAndSwap` operation, so it is atomic, consistent and will do completely fine.
`getAndUpdate` is also fine with one exception that it will reduce throughput of contended operations on `state` for Java 8 (starting from Java 9 it is fine), for additional details, please read the doc comment on `updateStateWithBackOff` method.

Reply:
reimplemented drain using compareAndSet

please check if this impl is ok for you

I also added a unit test for an drain edge case